File: C#/Leetcode/Patterns/Sliding Window/FixedWindowSize/MaximumSumOfDistinctSubarraysWithLengthK.cs
public class Solution {
    // Leetcode 2461 - https://leetcode.com/problems/maximum-sum-of-distinct-subarrays-with-length-k/description/
    // Submission Detail - https://leetcode.com/problems/maximum-sum-of-distinct-subarrays-with-length-k/submissions/1541459500
    // Fixed Window Size
    // nums = [3, 2, 2, 3, 4, 6, 7, 7, -1]
    // k = 4
    // Output = 13

    // Tx = O(n)
    // Sx = O(n)
    public long MaximumSubarraySum(int[] nums, int k) {
        long currentSum = 0, maxSum = 0;
        int start = 0, length = nums.Length;
        IDictionary<int, int> numFrequency = new Dictionary<int, int>(length);

        for(int end = 0; end < length; end++) {
            currentSum += nums[end];
            if(!numFrequency.ContainsKey(nums[end]))
                numFrequency.Add(nums[end], 0);
            numFrequency[nums[end]]++;

            // Window Size is k
            if(end-start+1 == k) {
                // Distinct numbers in the window found
                if(numFrequency.Keys.Count == k) {
                    maxSum = Math.Max(maxSum, currentSum);
                }
                currentSum -= nums[start];
                numFrequency[nums[start]]--;
                if(numFrequency[nums[start]] == 0)
                    numFrequency.Remove(nums[start]);
                start++;
            }
        }

        return maxSum;
    }
}

File: Java/SlidingWindow/FixedWindow/MaximumSumofDistinctSubarraysWithLengthK.java
package SlidingWindow.FixedWindow;

import java.util.HashMap;
import java.util.Map;

public class MaximumSumofDistinctSubarraysWithLengthK {
    // Leetcode 2461: https://leetcode.com/problems/maximum-sum-of-distinct-subarrays-with-length-k/
    // Submission Detail: https://leetcode.com/problems/maximum-sum-of-distinct-subarrays-with-length-k/submissions/1625794122

    //Input: nums = [3, 2, 2, 3, 4, 6, 7, 7, -1], k = 4
    //Output: 20
    public long maximumSubarraySum(int[] nums, int k) {
        int start = 0;
        long maxSum = 0, currentSum = 0;
        Map<Integer, Integer> numCount = new HashMap<>();
        for(int end = 0; end < nums.length; end++) {
            currentSum += nums[end];
            int endIndexCount = numCount.getOrDefault(nums[end], 0);
            numCount.put(nums[end], endIndexCount+1);

            if(end-start+1 == k) {
                if(numCount.size() == k) {
                    maxSum = Math.max(maxSum, currentSum);
                }
                currentSum -= nums[start];

                int startIndexCount = numCount.get(nums[start]);
                if(startIndexCount <= 1)
                    numCount.remove(nums[start]);
                else
                    numCount.put(nums[start], startIndexCount-1);

                start++;
            }
        }

        return maxSum;
    }
}
